internals/overlord/checkstate/manager.go

```diff
@@ -143,12 +268,42 @@ func newChecker(config *plan.Check, p *plan.Plan) checker {
 // Checks returns the list of currently-configured checks and their status,
 // ordered by name.
 func (m *CheckManager) Checks() ([]*CheckInfo, error) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// Populate name, number of failures, and change ID from state.
+	var infos []*CheckInfo
+	for _, change := range m.state.Changes() {
+		switch change.Kind() {
+		case performCheckKind, recoverCheckKind:
+			if change.IsReady() {
+				// Skip check changes that have finished already.
+				continue
+			}
```

Comment:
If I read this correctly, we are saying that running changes represents the active checks, and each running check holds the latest truth about its check state. I actually like this - it was easier for me to follow the flow of data, and ownership in general. Just a quick question now ... Is there an opportunity for a query (API) to happen exactly at a point in time between two changes (where the one change become ready - ```error``` or ```done```), and  the next change has not yet been committed? I think it boils down to when in the state engine the ```changeStatusChanged``` is called from. If while the state engine has its lock, the change transitions to ```Error``` / ```Done``` and while having the lock it calls the change callback, if feels like everything works perfectly for me.

Reply:
Yeah, I liked having all the data in the change/task too. I traced that through the state engine and yes, `changeStatusChanged` is called with the state locked throughout all this. [`TaskRunner.run` calls `SetStatus`](https://github.com/canonical/pebble/blob/18c93cbc280de20c9e21d323fe3a8bbb340777a2/internals/overlord/state/taskrunner.go#L268-L297) which updates the status from Doing to Done (or Error) and `SetStatus` calls the status-changed handler.

However, I had to bypass this "all the data in the change/task" due to the `GET /v1/health` endpoint requiring that the state lock isn't held. Saving modified state is relatively slow (due to the JSON serialisation and fsync), and in certain deployments when the machine was under load it would take too long to respond ([see details](https://github.com/canonical/pebble/pull/369)). We have a task on next cycle's roadmap to improve the speed/operation of state saving, but that's longer term.

In the meantime, I had to store a side map `CheckManager.health` with a mutex around it for the Health endpoint. I'm beginning to think this is a mistake, and I should remove the new parallel `Health` method and just have a `CheckManager.checks` map again, but it would be a `map[string]CheckInfo` instead of that funky `checkData` struct. All the operational logic and plan updates would still be done via changes. So I think I'll make them change yet, and I think it'll be simpler.

Reply:
I've now reverted the new `Health()` method and am using `CheckManager.checks map[string]CheckInfo` as described above.